crypto/src/serde_helpers.rs

```diff
@@ -0,0 +1,151 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::anyhow;
+use base64ct::Encoding as _;
+use blst::min_sig as blst;
+use schemars::JsonSchema;
+use serde::de::{Deserializer, Error};
+use serde::ser::Serializer;
+use serde::Deserialize;
+use serde::Serialize;
+use serde_with::{Bytes, DeserializeAs, SerializeAs};
+use std::fmt::Debug;
+
+fn to_custom_error<'de, D, E>(e: E) -> D::Error
+where
+    E: Debug,
+    D: Deserializer<'de>,
+{
+    D::Error::custom(format!("byte deserialization failed, cause by: {:?}", e))
+}
+
+pub struct BlsSignature;
+
+impl SerializeAs<blst::Signature> for BlsSignature {
+    fn serialize_as<S>(source: &blst::Signature, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        if serializer.is_human_readable() {
+            // Serialise to Base64 encoded String
+            Base64::encode(source.to_bytes()).serialize(serializer)
+        } else {
+            // Serialise to Bytes
+            Bytes::serialize_as(&source.serialize(), serializer)
+        }
+    }
+}
+
+impl<'de> DeserializeAs<'de, blst::Signature> for BlsSignature {
+    fn deserialize_as<D>(deserializer: D) -> Result<blst::Signature, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let bytes = if deserializer.is_human_readable() {
+            let s = String::deserialize(deserializer)?;
+            Base64::decode(&s).map_err(to_custom_error::<'de, D, _>)?
+        } else {
+            Bytes::deserialize_as(deserializer)?
+        };
+        blst::Signature::deserialize(&bytes).map_err(to_custom_error::<'de, D, _>)
+    }
+}
+
+pub struct Ed25519Signature;
+
+impl SerializeAs<ed25519_dalek::Signature> for Ed25519Signature {
+    fn serialize_as<S>(source: &ed25519_dalek::Signature, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        if serializer.is_human_readable() {
+            // Serialise to Base64 encoded String
+            Base64::encode(source.to_bytes()).serialize(serializer)
+        } else {
+            // Serialise to Bytes
+            Bytes::serialize_as(&source.to_bytes(), serializer)
+        }
+    }
+}
+
+impl<'de> DeserializeAs<'de, ed25519_dalek::Signature> for Ed25519Signature {
+    fn deserialize_as<D>(deserializer: D) -> Result<ed25519_dalek::Signature, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let bytes = if deserializer.is_human_readable() {
+            let s = String::deserialize(deserializer)?;
+            Base64::decode(&s).map_err(to_custom_error::<'de, D, _>)?
+        } else {
+            Bytes::deserialize_as(deserializer)?
+        };
+        ed25519_dalek::Signature::from_bytes(&bytes).map_err(to_custom_error::<'de, D, _>)
+    }
+}
+
+pub trait Encoding {
```

Comment:
I'm confused as to why we need to re-import all of this, which IIUC is essentially a copy of the base64 crate with a slightly different API: what am I missing?

Reply:
Removed the reimport and just used native functions - I ported this over from the Sui codebase and it was useful there because it shortened a lot of other code. No real need here.